src/patterns/scheduled-ecs-task.ts

```diff
@@ -0,0 +1,255 @@
+    repositoryName: app,
+  });
+ * const containerConfiguration = {
+ *     repository: Repository.fromRepositoryArn("<repository arn>"),
+ *     version: '1'
+ * }
+ * ```
+ */
+
+type RepositoryContainer = {
+  repository: IRepository;
+  version: string;
+  type: "repository";
+};
+
+type RegistryContainer = {
+  id?: string;
+  type: "registry";
+};
+
+type ContainerConfiguration = RepositoryContainer | RegistryContainer;
+
+type GuEcsTaskMonitoringProps = { snsTopicArn: string; noMonitoring: false };
+
+/**
+ * Configuration options for the [[`GuScheduledEcsTask`]] pattern.
+ *
+ * The `schedule` property determines when your task is invoked. For example, to invoke
+ * the lambda every 5 minutes, use:
+ * ```typescript
+ * import { Schedule } from "@aws-cdk/aws-events";
+ * import { Duration } from "@aws-cdk/core";
+ *
+ * const props = {
+ *   // Other props here
+ *   schedule: Schedule.rate(Duration.minutes(5)),
+ * }
+ * ```
+ *
+ * To invoke the task every weekday at 8am, use:
+ * ```
+ * import { Schedule } from "@aws-cdk/aws-events";
+ *
+ * const props = {
+ *   // Other props here
+ *   schedule: Schedule.expression("cron(0 8 ? * MON-FRI *)"),
+ * }
+ * ```
+ *
+ * See [[`ContainerConfiguration`]] for details of how to configure the container used to run the task.
+ *
+ * `taskTimeoutInMinutes` does what is says on the tin. The default timeout is 15 minutes.
+ *
+ * The `taskCommand` prop allows you to specify a command to run when the task starts (this can also be done via a CMD statement in
+ * your Dockerfile). For example:
+ *
+ * const props = {
+ *    //other props
+ *    taskCommand: `aws s3 cp s3://${distbucket}/${stack}/${stage}/${app}/task.sh . && ./task.sh
+    }
+ *
+ * It is advisable to configure alarms for when the job fails/times out. To do this specify the alarmSnsTopicArn prop.
+ *
+ * `customTaskPolicies` allows your task to interact with other AWS services. By default a task
+ * will have read access to the distribution bucket for your account.
+ *
+ * You can specify security groups to apply to task using the `securityGroups` prop.
+ *
+ * You can also set the memory and cpu units for your task. See https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#task_size
+ * for further details.
+ *
+ */
+export interface GuScheduledEcsTaskProps extends Identity {
+  vpc: IVpc;
+  schedule: Schedule;
+  containerConfiguration: ContainerConfiguration;
+  taskTimeoutInMinutes?: number;
+  cpu?: number;
+  memory?: number;
+  taskCommand?: string;
+  monitoringConfiguration: NoMonitoring | GuEcsTaskMonitoringProps;
+  securityGroups?: ISecurityGroup[];
+  customTaskPolicies?: PolicyStatement[];
+}
+
+/**
+ * Containers can be specified either via a repository and version or by container id (in which case it will
+ * be fetched from docker hub).
+ */
+const getContainer = (config: ContainerConfiguration) => {
+  if (config.type === "repository") {
```

Comment:
I think a [user defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) (essentially duck typing) would be more idomatic here.

Reply:
Thanks Akash, I attempted this here https://github.com/guardian/cdk/pull/791/commits/dfe51f534e9e61e4de69888e33056948f8dd16d3 but vscode is complaining about `(config as RepositoryContainer).version !== undefined` being an unncessary check. Would you have implemented it differently?

Reply:
Hmm, I think this is because `ContainerConfiguration` is a union type? If you make `config` an `any` type it should be happier (something similar [here](https://github.com/guardian/cdk/blob/1f1735d3eea11686e24cccb6ee6bdc2aef661362/src/types/migrating.ts#L14-L16)).

Having said that, [these docs](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) follow the same pattern as before (switching on  the `type` field). I'm not sure what's best anymore! I'd go for whatever is easiest, so probably the switching. Sorry!

Reply:
Thanks! I'm going to be lazy as at least it's quite readable - those docs were the ones I followed originally

Reply:
Sorry I did reply to this but I think my internet broke when I hit comment. Thanks for the feedback! I decided to leave it as is - isSomething is a more readable notation but in this instance where we're only checking it once doing the type == thing seems ok to me - or at least that's how I justified my laziness